[danielleloh] iP

danielleloh opened the pull request:
# Scribbles
Scribbles helps you to track and manage the many tasks in your life! ✔️ 

It is:
- Text-based,
- Interactive,
- Simple and neat!
---
How to use Scribbles:
1. Click [here](https://github.com/danielleloh/ip) to download it
2. Launch it
3. And start using it!

The best thing about Scribbles is, it's **FREE** for all!

---
Features:
- [x] Manages your to-dos
- [x] Manages your deadlines
- [x] Manages your events
- [x] And more features coming up!

---
If you are learning Java, you can use it to practice your Java too! Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```
---
Remember,
> “Your mind is for having ideas, not holding them.” – David Allen

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

danielleloh committed:
Rename, Greet, and Exit

danielleloh committed:
Edited to echo commands entered by user, exits when user enters bye

danielleloh committed:
Added a list to store tasks, and list items in list

danielleloh committed:
Added mark and unmark feature

danielleloh committed:
Added task types: Todos, Deadlines, Events.

danielleloh committed:
Edited System.out.println when adding Todos, Deadlines, Events to list

danielleloh committed:
Added automated text UI testing

danielleloh committed:
Added error messages for invalid inputs.

danielleloh committed:
Add support for deleting tasks from the list.

danielleloh committed:
Separate classes Task, Todo, Deadline and Event from main method

For better organization of classes from the main method

danielleloh committed:
Rename file Duke.java to Scribbles.java

Rename file to match with new chatbot name

danielleloh committed:
Scribbles.java: Add loadFileData method

Data from the hard disk needs to be loaded to program when chatbot starts up.

Method is called when main() is ran and the data from hard disk is loaded into the chatbot.

danielleloh committed:
Scribbles.java: Add saveFileData method

Update the hard disk whenever changes to tasks are made

This saves the data to the hard disk such that when main() is ran again, it is able to load data from the previous use of the chatbot.

danielleloh committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Scribbles.java: Add saveFileData method
  Scribbles.java: Add loadFileData method
  Rename file Duke.java to Scribbles.java
  Separate classes Task, Todo, Deadline and Event from main method

danielleloh committed:
Scribbles.java: Add file and directory check

Checks if the file path used by the chatbot exists in the user's computer. If it does not, it will create the directory and the file in the user's computer to save data.

danielleloh committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Scribbles.java: Add file and directory check

danielleloh committed:
Event.java: Change the start and end type to LocalDateTime

Scribbles can now recognize the start and end in Event as a date and time.

Prints stored start date/time and end date/time in the format "dd/MM/yyyy HHmm" such that it is readable to users of the chatbot.

danielleloh committed:
Scribbles.java: Edit all methods affected by the change in type of end and start of Event from String to LocalDateTime

Scribbles now recognize start and end of Event as a date/time.

Edit,
* addEvent method to convert String input of user in format "dd/MM/yyyy HHmm" to date/time of type LocalDateTime
* addEvent method to print error message if date/time is in incorrect format
* loadFileData method to add saved date/time of Event to LocalDateTime before adding to taskList
* loadFileData method to print error message if date/time for Event stored are of incorrect format
* saveFileData method to save to file date/time of Event as String in the format "dd/MM/yyyy HHmm" from type LocalDateTime

danielleloh committed:
Deadline.java: Edit type of By to be LocalDateTime

Chatbot is able to recognise By as a LocalDateTime

Edit,
* Type of By from String to LocalDateTime
* getBy() method to return By in LocalDateTime
* Add getByString() method to return By in String format "dd/MM/yyyy HHmm"
* toString() method to return By in String format "dd/MM/yyyy HHmm" that is readable to user

danielleloh committed:
Scribbles.java: Edit methods that are affected by change of type of By to LocalDateTime in Deadline.java

Change of type of By from String to LocalDateTime will cause bugs in methods that are affected.

Edit,
* addDeadline method to read deadline input in String format "dd/MM/yyyy HHmm" and store it as a LocalDateTime in taskList
* addDeadline method to print a warning if input is not in String format "dd/MM/yyyy HHmm"
* loadFileData to add deadlines stored in hard disk in the String format "dd/MM/yyyy HHmm" to taskList as LocalDateTime
* saveFileData to save Deadline to harddisk of type LocalDateTime to String format "dd/MM/yyyy HHmm"

danielleloh committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Scribbles.java: Edit methods that are affected by change of type of By to LocalDateTime in Deadline.java
  Deadline.java: Edit type of By to be LocalDateTime
  Scribbles.java: Edit all methods affected by the change in type of end and start of Event from String to LocalDateTime
  Event.java: Change the start and end type to LocalDateTime

danielleloh committed:
Add new classes Ui.java, TaskList.java, Storage.java and Parser.java.

This makes the code more OOP.

* class Ui.java: deals with interactions with the user
* class Storage.java: deals with loading tasks from the file and saving tasks in the file
* class Parser.java: deals with making sense of the user command
* class TaskList.java: contains the task list

danielleloh committed:
Merge branch 'branch-OOP-Increment'

* branch-OOP-Increment:
  Add new classes Ui.java, TaskList.java, Storage.java and Parser.java.

danielleloh committed:
Create package for classes.

Organise classes into suitable packages.

Create,
* task package for Task.java, Todo.java, Deadline.java, Event.java
* parser package for Parser.java
* storage package for Storage.java
* ui package for Ui.java

Add import statements to affect classes to allow use of classes in other packages.

danielleloh committed:
Merge branch 'branch-Package-Increment'

* branch-Package-Increment:
  Create package for classes.

danielleloh committed:
Merge commit 'a6f73244a20539e20c2189973ebf597df34ef631'

* commit 'a6f73244a20539e20c2189973ebf597df34ef631':
  Bump gradle and lib version
  Add Gradle support
  docs/README.md: Tweak document template

danielleloh committed:
ScribblesTest.java: Create JUnit test for parseInput(String input) method in Scribbles.java.

Automate the testing of parseInput for different input to ensure the right output.

Wrote test cases for all possible inputs.

danielleloh committed:
ParserTest.java: Create JUnit tests for isMissingDeadlineInformation() method in Parser class

Automate the testing of isMissingDeadlineInformation for different inputs to ensure the right output.
It should return true if any inputs for deadlines are missing relevant information, and true otherwise, or throw IndexOutOfBoundsException

Wrote test cases for all possible inputs.

danielleloh committed:
Merge branch 'branch-JUnit-Increment'

* branch-JUnit-Increment:
  ParserTest.java: Create JUnit tests for isMissingDeadlineInformation() method in Parser class
  ScribblesTest.java: Create JUnit test for parseInput(String input) method in Scribbles.java.

danielleloh committed:
Add JavaDoc comments to code.

Explain the use of the different methods and classes.

danielleloh committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add JavaDoc comments to code.

danielleloh committed:
Edit code to comply with coding standards.

Complying with coding standards will help to improve code quality.

Edit,
* boolean methods and variables to sound like booleans — completed to isCompleted, parsedInput to isParsedInput
* indentation for switch statements — remove the indentation in front of case clauses

danielleloh committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Edit code to comply with coding standards.

danielleloh committed:
Parser.java: Change split method input from ("\\s+") to ("\\s+", 2)

The error was causing to-do tasks description with a space in between to only store the first word.

danielleloh committed:
Add feature that give users a way to search for tasks using a keyword.

Users can easily search for tasks with a keyword instead of looking through the whole list of tasks.

Add,
* getFindKeyword() method to Parser.java that returns the keyword in the user's input
* find case clause for the actions to perform after a find command
* new instruction under invalidInputMessage to direct users on how to use find command
* taskWithKeyword method Ui.java class that lists all the tasks with the keyword
* missingKeywordMessage() method that prints an error message if users do not include a keyword in their find command

danielleloh committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add feature that give users a way to search for tasks using a keyword.
  Parser.java: Change split method input from ("\\s+") to ("\\s+", 2)

danielleloh committed:
fix: allow todo tasks to accept a description

Split string in getTodoDescription() split the string wrongly causing
it to throw an IndexOutOfBoundsException all the time.

Change ("\\s+, 2") to ("\\s+", 2).

danielleloh committed:
fix: edit line spacing of scribbles messages

some replies include a line spacing after and some do not which
makes it seem messy.

danielleloh committed:
fix: add checkstyle feature and edit checkstyle errors

danielleloh committed:
Merge branch 'branch-CheckStyle'

* branch-CheckStyle:
  fix: add checkstyle feature and edit checkstyle errors

matthewken19 commented on src/main/java/scribbles/Scribbles.java:
```diff
@@ -0,0 +1,227 @@
+                    String description = parsedInput.getEventDescription();
+                    LocalDateTime start = parsedInput.getStartDateTime();
+                    LocalDateTime end = parsedInput.getEndDateTime();
+                    taskList.addTask(new Event(description, false, start, end));
+                    ui.confirmTaskAddition(taskList);
+                    storage.saveFileData(taskList);
+                } catch (IndexOutOfBoundsException e) {
+                    ui.taskMissingInformationMessage();
+                } catch (DateTimeParseException e) {
+                    ui.wrongDateTimeFormatMessage();
+                } catch (FileNotFoundException e) {
+                    ui.fileNotFoundMessage();
+                }
+            }
+            break;
+        case "delete":
+            try {
+                int index = parsedInput.getIndex();
+                String taskRemoved = taskList.get(index - 1).toString();
+                taskList.deleteTask(index - 1);
+                ui.taskDeletionMessage(taskRemoved, taskList);
+                storage.saveFileData(taskList);
+            } catch (IndexOutOfBoundsException e) {
+                ui.invalidIndexMessage(taskList);
+            } catch (FileNotFoundException e) {
+                ui.fileNotFoundMessage();
+            }
+            break;
+        case "find":
+            try {
+                String keyword = parsedInput.getFindKeyword();
+                ui.tasksWithKeyword(keyword, taskList);
+            } catch (IndexOutOfBoundsException e) {
+                ui.missingKeywordMessage();
+            }
+            break;
+        default:
+            ui.invalidInputMessage();
+        }
+        return false;
```
I think you can create your own exception and catch it at the end of the switch case statement so it will only handle your custom exception instead of handling all exceptions in every case.

matthewken19 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,124 @@
+                case "T":
+                    taskList.addTask(new Todo(description, isCompleted));
+                    break;
+                case "D":
+                    String deadlineString = tokens[3].trim();
+                    try {
+                        DateTimeFormatter dateTimeFormat = DateTimeFormatter.ofPattern("dd/MM/yyyy HHmm");
+                        LocalDateTime deadline = LocalDateTime.parse(deadlineString, dateTimeFormat);
+                        taskList.addTask(new Deadline(description, isCompleted, deadline));
+                    } catch (DateTimeParseException e) {
+                        System.out.println("Your deadline, \"" + description + "\" stored in file has incorrect "
+                                + "date/time format and cannot be loaded into task list. ");
+                    }
```
good handling of DateTimeParseException

danielleloh replied:
Thank you :)

matthewken19 commented on src/main/java/storage/Storage.java:
```diff
@@ -0,0 +1,124 @@
+            BufferedWriter writer = new BufferedWriter(new FileWriter(filePath));
+            for (int i = 0; i < taskList.size(); i++) {
+                Task task = taskList.get(i);
+                if (task instanceof Todo) {
+                    writer.write("T | " + (task.isCompleted() ? "1" : "0") + " | " + task.getDescription());
+                }
+                if (task instanceof Deadline) {
+                    writer.write("D | " + (task.isCompleted() ? "1" : "0") + " | " + task.getDescription()
+                            + " | " + ((Deadline) task).getByString());
+                }
+                if (task instanceof Event) {
+                    writer.write("E | " + (task.isCompleted() ? "1" : "0") + " | " + task.getDescription()
+                            + " | " + ((Event) task).getStartString() + " | " + ((Event) task).getEndString());
+                }
```
I think you can include this method in the classes instead and use polymorphism from the Task class method

danielleloh replied:
Thank you for your advice :) I will try implementing that

matthewken19 commented on src/test/java/parser/ParserTest.java:
```diff
@@ -0,0 +1,31 @@
+public class ParserTest {
+
+    @Test
+    public void isMissingDeadlineInformationTest() {
```
I like how thorough you are with test cases! However I think naming convention should be something like: 
isMissingDeadlineInformation_correctFormat_success() or isMissingDeadlineInformation_incorrectFormat_exceptionThrown() 
to show unitBeingTested_descriptionOfTestInputs_expectedOutcome in the test name

danielleloh replied:
Thank you for the feedback:)

matthewken19 reviewed (commented):
Overall, I like the OOP abstraction! LGTM, but just take note of some changes that I have proposed that might help!

danielleloh committed:
feat: add basic gui to scribbles

Use the JavaFX technology to implement the GUI.

danielleloh committed:
Merge branch 'branch-GUI'

* branch-GUI:
  feat: add basic gui to scribbles

brennalaurentan commented on src/main/java/task/Task.java:
```diff
@@ -0,0 +1,54 @@
+     * @return User-readable task in String format
+     */
+    public String toString() {
+        if (this.isCompleted) {
```
Not a coding standard violation and this is fine, but can consider having a `completionIcon()` method which returns "[ ]" or "[X]" depending on the boolean isCompleted status instead of statically using if & else to determine which one to print

then this method can just be `return this.completionIcon() + " " + this.description`

will be neater if you decide to add more components to this `toString()` method in the future as you have more control over the individual components :)

danielleloh replied:
Thank you for your advice! It really does seem neater if I implement it that way. I will try :)

brennalaurentan commented on src/test/java/scribbles/ScribblesTest.java:
```diff
@@ -0,0 +1,49 @@
+public class ScribblesTest {
+
+    @Test
+    public void parseInputTest() {
```
Good use of JUnit to implement test cases! Comprehensive testing of software 👍

danielleloh replied:
Thank you :)

brennalaurentan commented on src/main/java/task/Deadline.java:
```diff
@@ -0,0 +1,38 @@
+import java.time.format.DateTimeFormatter;
+
+/**
+ * This class represents a deadline
```
Maybe can provide a little more detail in the comment header for the class to indicate how Todo, Deadline and Event tasks differ? Such as mentioning how Deadline has a due date and Event has from and to dates

danielleloh replied:
Thank you for your feedback! What you say makes sense and I will take it into consideration and edit it to be clearer :)

brennalaurentan commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,51 @@
+import java.time.format.DateTimeFormatter;
+
+/**
+ * This class represents an event.
```
Maybe can provide a little more detail in the comment header for the class to indicate how Todo, Deadline and Event tasks differ? Such as mentioning how Deadline has a due date and Event has from and to dates

danielleloh replied:
Thank you for your feedback! What you say makes sense and I will take it into consideration and edit it to be clearer :)

brennalaurentan reviewed (commented):
Nice implementation of your own task manager! No violations found, but just a few suggestions here and there.

danielleloh committed:
feat: add default greeting message in GUI when it is launched

danielleloh committed:
Merge branch 'branch-GUI'

* branch-GUI:
  feat: add default greeting message in GUI when it is launched

danielleloh committed:
fix: change icon for GUI and edit some view features

danielleloh committed:
fix: edit javadocs

danielleloh committed:
fix: change minimum height of dialog box and edit gui aesthetics

messages that exceed height indicated would automatically be
truncated with "...".

update the dialog box minimum height such that the minimum height is
USE_PREF_SIZE.

danielleloh committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  fix: change minimum height of dialog box and edit gui aesthetics

danielleloh committed:
fix: change marking completed icons to be constant strings

changing the icons to constant string help to avoid magic literals as
it makes it clearer what they icons represent

add private static final String COMPLETED_ICON and INCOMPLETED_ICON

danielleloh committed:
fix: edit help message to be more informative and fix typos

danielleloh committed:
feat: add java assertions to code

java assertions help to document important assumptions that should hold
at various points of the code

danielleloh committed:
fix: change the dateTimeFormat of Parser class to a constant

same date time format is used throughout the class.
improves code quality by avoiding magic literals.

add private static final DATE_TIME_FORMAT.

danielleloh committed:
fix: split getResponse method and add another method respondToCommand

improve code quality by avoiding multiple levels of abstraction within
the getResponse method.

danielleloh committed:
Merge pull request #1 from danielleloh/branch-A-Assertions

feat: add java assertions to code

danielleloh committed:
Merge pull request #2 from danielleloh/branch-A-CodeQuality

feat: edit to improve code quality

danielleloh committed:
feat: add feature to allow users to sort the tasks in their task list

users may find it neater to sort their tasks for neater organisation

allowed sorting orders:
* date
* description
* completed first
* incompleted first
* type of task

danielleloh committed:
Merge branch 'branch-BCD-Extensions'

* branch-BCD-Extensions:
  feat: add feature to allow users to sort the tasks in their task list

danielleloh committed:
fix: update javadocs for new methods added

danielleloh committed:
fix: fix gui such that user inputs are aligned to the right

danielleloh committed:
fix: change mainClassName in build.gradle to set new entry point of
the application to scribbles.Launcher instead of scribbles.Scribbles

danielleloh committed:
feat: improve GUI to make it look more polished

add,
* chat boxes surrounding dialog of Scribbles and User
* different fonts to accommodate windows users and macbook users
* different display image to look better and clearer
* different colours to the chat box and text for dialog of Scribbles
and User

danielleloh committed:
feat: add quitting of program after user input "bye"

the chatbot only responds to bye initially but does not completely
exit and shut down the program.

add a 2 second delay after bye for the program to shut down.

danielleloh committed:
fix: disallow the input of event start date/time to be later than
event end date/time

it does not make sense for the start of the event to be later than the
end of the event

chatbot does not allow the addition of such event and prints a message
to inform users that the input start is later than the input end

danielleloh committed:
fix: change filepath of where there data is stored so that it will can
work when converted to a JAR file

danielleloh committed:
fix: edit warning message to be clearer when file date is corrupted

danielleloh committed:
fix: add javadocs for methods

danielleloh committed:
fix: edit name of directory for filepath for Scribbles

the name is edited to be more specific that it is linked to Scribbles
so that users know which file contains information for Scribbles.

danielleloh committed:
fix: add class CommandResponder to deal with responding to commands
made by the user

this resolves the very-long method, respondToCommand, which was
initially in the Scribbles class by further abstracting the method.
This also improves OOP as the response to the commands is dealt with in
a separate class.

add,
* new class CommandResponder
* methods for the different commands
* javadocs for the new methods in CommandResponder

danielleloh committed:
fix: update junit testing to accomodate new or removed methods

danielleloh committed:
feat: add image of Ui into docs folder

danielleloh committed:
fix: add response to "bye" and other input command into CommandResponder
folder

This improves OOP as it improve encapsulation as all response to
commands are not encapsulated in the CommandResponder class.

danielleloh committed:
feat: add about to user guide

introduces Scribbles to potential users

danielleloh committed:
fix: change ui preview image link to direct link

danielleloh committed:
fix: fix ui preview alignment and size

danielleloh committed:
feat: add table of contents for user guide

danielleloh committed:
fix: fix file formatting for user guide

danielleloh committed:
fix: edit clickable links to work and add in first version of
quick start instructions in user guide

danielleloh committed:
fix: update user guide to include more details on the available
commands and add a command summary

danielleloh committed:
fix: add blank lines to user guide

blank lines help to separate the content more visibly allowing for
better readability.

danielleloh committed:
fix: add blank lines to user guide to improve readability

danielleloh committed:
fix: fix format of README file

danielleloh committed:
fix: edit format of README file

danielleloh committed:
fix: fix format of table in user guide

danielleloh committed:
feat: add link to scribbles.jar release into user guide

danielleloh committed:
fix: update user guide to include more information on the data storage
of Scribbles.

danielleloh committed:
fix: correct typos in user guide

danielleloh committed:
fix: changed Ui preview image to be linked to my github docs Ui.png

danielleloh committed:
docs: update project README

danielleloh committed:
docs: update project README
